src/charm.py

```diff
@@ -56,12 +59,24 @@ def __init__(self, *args):
         self.grafana_source_provider = GrafanaSourceProvider(
             charm=self,
             refresh_event=self.on.loki_pebble_ready,
-            source_type="loki",
+            source_type=PEER,
             source_port=str(self._port),
         )
         self._loki_server = LokiServer()
         self._provide_loki()
-        self.loki_provider = LokiPushApiProvider(self)
+
+        self.ingress_per_unit = IngressPerUnitRequirer(
+            self, endpoint="ingress-per-unit", port=self._port
+        )
+        external_url = urlparse(self._external_url)
+
+        self.loki_provider = LokiPushApiProvider(
+            self,
+            endpoint_address=external_url.hostname or "",
+            endpoint_port=external_url.port or self._port,
+            endpoint_schema=external_url.scheme,
+            endpoint_path=f"{external_url.path}/loki/api/v1/push",
```

Comment:
I believe Loki needs to be started with `-server.path-prefix=`, see https://stackoverflow.com/questions/62575304/how-to-configure-loki-to-accept-requests-with-subpath

Reply:
What do you think using `http_path_prefix` and place this setting in the config file instead of starting the sever with the `--server.path-prefix` flag?

See: https://grafana.com/docs/loki/latest/configuration/#server_config

Reply:
If it works, I have no particular preference one way or another. I am used to debug path issues with `ingress` using `kubectl exec ... -- ps aux`, but checking out the file works fine too.

Reply:
When a relation between Loki and Traefik is established the `server` section in Loki's config file is:

```yaml
server:
  http_listen_address: 0.0.0.0
  http_listen_port: 3100
  http_path_prefix: /cos-loki-0
``` 

![imagen](https://user-images.githubusercontent.com/939888/161143981-3b30437d-4a71-4a15-8374-0e13f7040f9c.png)


When there is not relation:

```yaml
server:
  http_listen_address: 0.0.0.0
  http_listen_port: 3100
  http_path_prefix: ''
```

![imagen](https://user-images.githubusercontent.com/939888/161144065-92521860-9896-4531-9441-7f265a81709d.png)